WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift

```diff
@@ -0,0 +1,237 @@
+                /// When we stop using proximity, we'll need more elaborate logic as we
+                /// won't be able to do that anymore
+                ///
+                self?.foundReader = cardReaders.first
+
+                guard let foundReaderID = self?.foundReader?.id else {
+                    DDLogWarn("onBeginSearch unexpectedly handled a nil found reader")
+                    self?.state = .endSearch
+                    return
+                }
+
+                /// If we know this reader, automatically connect to it
+                ///
+                let knownReaderIDs = self?.knownReaderIDs ?? []
+                if knownReaderIDs.contains(foundReaderID) {
+                    self?.state = .connectToReader
+                } else {
+                    self?.state = .foundReader
+                }
+            },
+            onError: { [weak self] error in
+                ServiceLocator.analytics.track(.cardReaderDiscoveryFailed, withError: error)
+                self?.state = .failed(error)
+            })
+
+        ServiceLocator.stores.dispatch(action)
+    }
+
+    /// A search is in progress
+    ///
+    func onSearching() {
+        alerts.scanningForReader(from: fromController, cancel: {
+            self.state = .endSearch
+        })
+    }
+
+    /// End the search for a card reader
+    ///
+    func onEndSearch() {
+        let action = CardPresentPaymentAction.cancelCardReaderDiscovery() {_ in
+            self.state = .idle
+        }
+        ServiceLocator.stores.dispatch(action)
+    }
+
+    /// A reader has been found
+    ///
+    func onFoundReader() {
+        guard foundReader != nil else {
+            DDLogWarn("onFoundReader unexpectedly called with no found reader")
+            self.state = .endSearch
+            return
+        }
+
+        guard let name = foundReader?.id else {
+            DDLogWarn("Found reader unexpectedly had no id (name)")
+            self.state = .endSearch
+            return
+        }
+
+        alerts.foundReader(
+            from: fromController,
+            name: name,
+            connect: {
+                self.state = .connectToReader
+            },
+            continueSearch: {
+                self.state = .searching
+            })
+    }
+
+    /// Connect to a card reader
+    ///
+    func onConnectToReader() {
+        guard let reader = foundReader else {
+            DDLogWarn("No found reader to connect to")
+            self.state = .endSearch
+            return
+        }
+
+        let action = CardPresentPaymentAction.connect(reader: reader) { [weak self] result in
+            switch result {
+            case .success(let reader):
+                self?.knownCardReadersProvider.rememberCardReader(cardReaderID: reader.id)
+                // If the reader does not have a battery, or the battery level is unknown, it will be nil
+                let properties = reader.batteryLevel
```

Comment:
I'm just curious here. Could this be `let properties = ["battery_level": reader.batteryLevel`? What is the benefit of passing through map?

Reply:
That would not give the same result. Given that the tracks implementation is done in Objective-C, I'm not even sure what would happen if we pass a dictionary with a `nil` value there 😰 

![Screen Shot 2021-08-04 at 12 48 18](https://user-images.githubusercontent.com/8739/128168625-480c904d-cd17-4038-b99b-52f71f10bdb8.png)